### ISSUE TITLE
podman 5.0.1 crashing with "fatal error: concurrent map writes"

### ISSUE THREAD

**Issue by kyrias** [containers/podman#22330]
### Issue Description

Since upgrading to podman 5 for our GitLab runner we've been experiencing podman crashing with a "fatal error: concurrent map writes" error occasionally while the system is under load.

### Steps to reproduce the issue

We don't have any simple reproducing case.  It happens regularly while the machine is running 100 gitlab CI jobs at once.

### Describe the results you received

Podman process crashes with "fatal error: concurrent map writes".

Full backtrace [attached](https://github.com/containers/podman/files/14934073/podman-crash.log).

### Describe the results you expected

Podman not crashing.

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.35.3
  cgroupControllers:
  - cpu
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: /usr/bin/conmon is owned by conmon 1:2.1.10-1
    path: /usr/bin/conmon
    version: 'conmon version 2.1.10, commit: 2dcd736e46ded79a53339462bc251694b150f870'
  cpuUtilization:
    idlePercent: 73.38
    systemPercent: 9.06
    userPercent: 17.57
  cpus: 96
  databaseBackend: sqlite
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  freeLocks: 1929
  hostname: elokon-gitlab-runner2
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 107
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 107
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 6.8.4-arch1-1
  linkmode: dynamic
  logDriver: journald
  memFree: 47246692352
  memTotal: 134602985472
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: Unknown
      path: /usr/local/lib/podman/aardvark-dns
      version: aardvark-dns 1.11.0-dev
    package: /usr/lib/podman/netavark is owned by netavark 1.10.3-1
    path: /usr/lib/podman/netavark
    version: netavark 1.10.3
  ociRuntime:
    name: crun
    package: /usr/bin/crun is owned by crun 1.14.4-1
    path: /usr/bin/crun
    version: |-
      crun version 1.14.4
      commit: a220ca661ce078f2c37b38c92e66cf66c012d9c1
      rundir: /run/user/107/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: /usr/bin/pasta is owned by passt 2024_04_05.954589b-1
    version: |
      pasta 2024_04_05.954589b
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user/107/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: /usr/bin/slirp4netns is owned by slirp4netns 1.2.3-1
    version: |-
      slirp4netns version 1.2.3
      commit: c22fde291bb35b354e6ca44d13be181c76a0a432
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.5
  swapFree: 66627563520
  swapTotal: 67301273600
  uptime: 2h 5m 7.00s (Approximately 0.08 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries: {}
store:
  configFile: /var/lib/gitlab-runner/.config/containers/storage.conf
  containerStore:
    number: 80
    paused: 0
    running: 6
    stopped: 74
  graphDriverName: btrfs
  graphOptions: {}
  graphRoot: /var/lib/gitlab-runner/.local/share/containers/storage
  graphRootAllocated: 1919342870528
  graphRootUsed: 733597417472
  graphStatus:
    Build Version: Btrfs v6.8
    Library Version: "102"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 3
  runRoot: /run/user/107/containers
  transientStore: false
  volumePath: /var/lib/gitlab-runner/.local/share/containers/storage/volumes
version:
  APIVersion: 5.0.1
  Built: 1712088128
  BuiltTime: Tue Apr  2 20:02:08 2024
  GitCommit: 946d055df324e4ed6c1e806b561af4740db4fea9-dirty
  GoVersion: go1.22.1
  Os: linux
  OsArch: linux/amd64
  Version: 5.0.1
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

**Comment by Luap99** [comment]
The stack trace points to https://github.com/containers/podman/blob/17f36a369035b4d5c27f416af9cb961ede807eed/pkg/api/handlers/compat/networks.go#L124

And that definitely writes the map (i.e. deletes a item), the code is broken as it should clone the map first otherwise it will change options in the backend which is not good. I guess I thought `options := network.Options` would clone the map but that is not the case. Easy enough to fix, I will open a PR.

This code was added in https://github.com/containers/podman/commit/52656887e1cd3476a18765aa81a7804fceb861b4 for v4.3 so I am surprised that you just now notice this in 5.0, likely some other change that triggered it.

**Comment by kyrias** [comment]
That was quick, thanks!

**Comment by Bitamarghzari** [comment]
عالی بود ممنون

**Comment by lolli42** [comment]
Hope its ok to chime in.
It seems we're affected by this as well, but with podman 4.9.3 / 4.9.4, a gitlab-runner instance, too.
Trace from line 153 in this version.
Is there any chance to have the fix merged to 4.9 and released?

**Comment by Luap99** [comment]
Not sure if it applies cleanly but if someone wants to do to the work to backport and tests then maybe.
We only support the latest release, 4.9 might get further security releases so it could get picked up in the next security release but no promises.

Any reason you cannot move to 5.0?

**Comment by lolli42** [comment]
Thanks!

I was unaware of podman's release strategy regarding older branches, thus the 4.9 release question.

I have no idea of go, but I can of course try to create a blind PR and see if it satisfies builds. If not, I'm probably out of the game.  We'll also try to build locally to test the fix with 4.9.

We just recently upgraded to ubuntu 24.04 hosts which packages podman 4.9.3 as "outer" podman, and combine with 4.9.4 as inner podman. There is no hard blocker to 5.0 (we didn't try yet), but if we can avoid putting additional time into building this major upgrade ourselfs, we'd try to keep 4.9 for now.

**Comment by Bitamarghzari** [comment]
Thanks

**Comment by Luap99** [comment]
Looks like it can be backported cleanly https://github.com/containers/podman/pull/22835. If this is merged we should get another 4.9 release soon with a security fix as well
